File: java/0054-spiral-matrix.java
class Solution {

    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> list = new ArrayList<>();
        int rb = 0;
        int re = matrix.length - 1;
        int cb = 0;
        int ce = matrix[0].length - 1;

        while (rb <= re && cb <= ce) {
            for (int j = cb; j <= ce; j++) {
                list.add(matrix[rb][j]);
            }
            rb++;

            for (int i = rb; i <= re; i++) {
                list.add(matrix[i][ce]);
            }
            ce--;

            if (rb <= re) {
                for (int j = ce; j >= cb; j--) {
                    list.add(matrix[re][j]);
                }
            }
            re--;

            if (cb <= ce) {
                for (int i = re; i >= rb; i--) {
                    list.add(matrix[i][cb]);
                }
            }
            cb++;
        }

        return list;
    }
}

File: swift/0054-spiral-matrix.swift
/**
 * Question Link: https://leetcode.com/problems/spiral-matrix/
 */

 class Solution {
    func spiralOrder(_ matrix: [[Int]]) -> [Int] {
        var res = [Int]()
        var l = 0
        var r = matrix[0].count
        var t = 0
        var b = matrix.count
        while l < r && t < b {
            for i in l..<r {
                res.append(matrix[t][i])
            }
            t += 1
            for i in t..<b {
                res.append(matrix[i][r - 1])
            }
            r -= 1
            if !(l < r && t < b) {
                break
            }
            for i in stride(from: r - 1, to: l - 1, by: -1) {
                res.append(matrix[b - 1][i])
            }
            b -= 1
            for i in stride(from: b - 1, to: t - 1, by: -1) {
                res.append(matrix[i][l])
            }
            l += 1
        }
        return res
    }
}